### PR TITLE
Correct check for write index and increment generation on all DS backing index operations

### PR DESCRIPTION
**Pull request by danhermann**
This corrects one instance where the check for the data stream's write index was using the generation rather than the number of backing indices. Often the two are the same, but they can get out of sync in a variety of situations such as after indices reach a delete step in ILM, snapshot restores, etc. In such a case, the `replaceBackingIndex` could have mistakenly allowed the replacement of the data stream's current write index.

Also, the generation should be incremented for any action that modifies backing indices so the `removeBackingIndex` and `replaceBackingIndex` methods were updated to increment the generation.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java [hunk 1]
```diff
@@ -290,18 +300,28 @@ public DataStream replaceBackingIndex(Index existingBackingIndex, Index newBacki
                 String.format(Locale.ROOT, "index [%s] is not part of data stream [%s]", existingBackingIndex.getName(), name)
             );
         }
-        if (generation == (backingIndexPosition + 1)) {
+        if (indices.size() == (backingIndexPosition + 1)) {
```
Comment (danhermann) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:303]: @martijnvg, when you have a minute, could you review this one? The main change is in this line though there are a couple others that I mentioned in the PR description.

**Comment by danhermann** [comment]
I will note that this includes changes that increment the generation of the data stream on operations other than rollover which could introduce what might be perceived as "gaps" in the sequence of backing indices. E.g., `.ds-foo-2021-01-01-000001` to `.ds-foo-2021-01-01-000003` if a "modify data stream" operation occurred before the rollover event. I don't consider that desirable, but it also seems appropriate that a data stream's generation should be incremented if its backing indices were changed by a "modify data stream" action.

**Review by martijnvg — APPROVED** [review]
Thanks @danhermann, LGTM 👍

**Comment by martijnvg** [comment]
>  I don't consider that desirable, but it also seems appropriate that a data stream's generation should be incremented if its backing indices were changed by a "modify data stream" action.

I think that an update (in any sense) should be made visible by incrementing the generation. So this makes sense to me.

**Comment by danhermann** [comment]
Thanks for reviewing and considering the question above, @martijnvg!

**Pull request merged**